vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+        MULTILABEL::labels preds;
+        if (b.oas ==false){
+            mc = ec.l.multi;
+            save_multi_pred = ec.pred.multiclass;
+        }
+        else{
+            multilabels = ec.l.multilabels;
+            preds = ec.pred.multilabels;
+        }
+	ec.l.simple = {FLT_MAX, 1., 0.0};
+        while(b.nodes[cn].internal != -1){
+            base.predict(ec, b.nodes[cn].base_router);
+            float prediction = ec.pred.scalar;
+            cn = prediction < 0 ? b.nodes[cn].left : b.nodes[cn].right;
+        }
+        
+        if(b.oas == false){
+            ec.l.multi = mc; 
+            ec.pred.multiclass = save_multi_pred;
+        }
+        else{
+            ec.pred.multilabels = preds;
+            ec.l.multilabels = multilabels;
+        }
+
+        //get to leaf now:
+	    int64_t closest_ec = 0;
+        float reward = 0.f;
+        closest_ec = pick_nearest(b, base, cn, ec); //no randomness for picking example.
+        if (b.oas == false){
+            if ((closest_ec != -1) && (b.examples[closest_ec]->l.multi.label == ec.l.multi.label))
+                reward = 1.f;
+        }
+        else{
+            if (closest_ec != -1)
+                reward = F1_score_for_two_examples(ec, *b.examples[closest_ec]);
+        }
+        b.total_num_queires++;
+
+        if (b.learn_at_leaf == true && closest_ec != -1){
+        	float score = normalized_linear_prod(b, &ec, b.examples[closest_ec]);
+        	example* kprod_ec = &calloc_or_throw<example>();
+        	diag_kronecker_product_test(ec, *b.examples[closest_ec], *kprod_ec, b.oas);
+         	kprod_ec->l.simple = {reward, 1.f, -score};
+	    	kprod_ec->weight = weight;
+            base.learn(*kprod_ec, b.max_routers);
+            free_example(kprod_ec);
+        }
+
+        if (b.oas == true)
+            train_one_against_some_at_leaf(b,base, cn,ec); ///learn the infoerence procedure anyway
+
+        return reward;
+    }
+
+    void learn_at_leaf_random(memory_tree& b, single_learner& base, const uint32_t& leaf_id, example& ec, const float& weight){
+        b.total_num_queires ++;
+        int32_t ec_id = -1;
+        float reward = 0.f;
+        if (b.nodes[leaf_id].examples_index.size() > 0){
+            uint32_t pos = uint32_t(merand48(b.all->random_state) * b.nodes[leaf_id].examples_index.size());
+			ec_id = b.nodes[leaf_id].examples_index[pos];
+        }
+        if (ec_id != -1){
+            if (b.examples[ec_id]->l.multi.label == ec.l.multi.label)
+                reward = 1.f;             
+            float score = normalized_linear_prod(b, &ec, b.examples[ec_id]);
+        	example* kprod_ec = &calloc_or_throw<example>();
+        	diag_kronecker_product_test(ec, *b.examples[ec_id], *kprod_ec, b.oas);
+         	kprod_ec->l.simple = {reward, 1.f, -score};
+	    	kprod_ec->weight = weight; //* b.nodes[leaf_id].examples_index.size();
+            base.learn(*kprod_ec, b.max_routers);
+            free_example(kprod_ec);
```

Comment:
Again, we want to avoid create/destroy.

Reply:
this is resolved now, via pre-allocated a space for the Kronecker product example, and free it at end of the entire learning procedure.